bqskit/compiler/passes/partitioning/scan.py

```diff
@@ -0,0 +1,180 @@
+    ) -> None:
+        """
+        Construct a SimplePartitioner.
+
+        Args:
+            block_size (int): Maximum size of partitioned blocks.
+                (Default: 3)
+
+            single_gate_score (int): When evaluating potential blocks,
+                use this number to score included single-qudit gates.
+                (Default: 1)
+
+            multi_gate_score (int): When evaluating potential blocks,
+                use this number to score included multi-qudit gates.
+                (Default: 1000)
+
+        Raises:
+            ValueError: If `block_size` is less than 2.
+        """
+
+        if not is_integer(block_size):
+            raise TypeError(
+                f'Expected integer for block_size, got {type(block_size)}.',
+            )
+
+        if block_size < 2:
+            raise ValueError(
+                f'Expected block_size to be greater than 2, got {block_size}.',
+            )
+
+        if not is_integer(single_gate_score):
+            raise TypeError(
+                'Expected integer for single_gate_score, '
+                f'got {type(single_gate_score)}.',
+            )
+
+        if not is_integer(multi_gate_score):
+            raise TypeError(
+                'Expected integer for multi_gate_score, '
+                f'got {type(multi_gate_score)}.',
+            )
+
+        self.block_size = block_size
+        self.single_gate_score = single_gate_score
+        self.multi_gate_score = multi_gate_score
+
+    def run(self, circuit: Circuit, data: dict[str, Any]) -> None:
+        """
+        Partition gates in a circuit into a series of CircuitGates.
+
+        Args:
+            circuit (Circuit): Circuit to be partitioned.
+
+            data (dict[str,Any]): Optional data unique to specific run.
+        """
+
+        if self.block_size > circuit.get_size():
+            _logger.warning(
+                'Configured block size is greater than circuit size; '
+                'blocking entire circuit.',
+            )
+            circuit.fold({
+                qudit_index: (0, circuit.get_num_cycles())
+                for qudit_index in range(circuit.get_size())
+            })
+            return
+
+        # If a MachineModel is provided in the data dict, it will be used.
+        # Otherwise all-to-all connectivity is assumed.
+        model = None
+
+        if 'machine_model' in data:
+            model = data['machine_model']
+
+        if (
+            not isinstance(model, MachineModel)
+            or model.num_qudits < circuit.get_size()
+        ):
+            _logger.warning(
+                'MachineModel not specified or invalid;'
+                ' defaulting to all-to-all.',
+            )
+            model = MachineModel(circuit.get_size())
+
+        # Find all connected, `block_size`-sized groups of qudits
+        # NOTE: This assumes circuit and topology qudit numbers are equal
```

Comment:
Is this note necessary? Does the partitioner still work if we're using a `MachineModel` that has more qudits than the circuit? This is the usual use case for doing the hybrid topology runs, and its seems that expanding the circuit so that it fits the size of the coupling graph might hurt performance (more locations to check).

Reply:
In this case, I think the loop over qudit groups would cause a crash since the circuit isn't explicitly expanded. We should probably log a warning or error here.